gobblin-tunnel/src/main/java/gobblin/tunnel/Tunnel.java

```
@@ -0,0 +1,178 @@
+ * 8. If established successfully, the Tunnel then simply relays bytes back and forth between the Gobblin extractor and
+ *    the target host via the intermediate proxy.)
+ * 7. When the Gobblin extractor (e.g. JDBC data source) is closed down, the Tunnel must be shut down as well.
+ *
+ * The Tunnel can accept as many connections as the JdbcExtractor opens. It uses NIO to minimize resource usage.
+ *
+ * @author navteniev@linkedin.com
+ * @author kkandekar@linkedin.com
+ */
+public class Tunnel {
+  public static final int NON_EXISTENT_PORT = -1;
+
+  private static final Logger LOG = LoggerFactory.getLogger(Tunnel.class);
+
+  private ServerSocketChannel server;
+  private Thread thread;
+  private final Config config;
+
+  private Tunnel(String remoteHost, int remotePort, String proxyHost, int proxyPort) {
+    config = new Config(remoteHost, remotePort, proxyHost, proxyPort);
+  }
+
+  private Tunnel open() throws IOException {
+    try {
+      server = ServerSocketChannel.open().bind(null);
+      startTunnelThread();
+
+      return this;
+    } catch (IOException ioe) {
+      LOG.error("Failed to open the tunnel", ioe);
+      throw ioe;
+    }
+  }
+
+  public int getPort() throws IOException {
+    SocketAddress localAddress = null;
+
+    try {
+      if (server != null && server.isOpen()) {
+        localAddress = server.getLocalAddress();
+      }
+      if (localAddress instanceof InetSocketAddress) {
+        return ((InetSocketAddress) localAddress).getPort();
+      }
+    } catch (IOException e) {
+      LOG.error("Failed to get tunnel port", e);
+      throw e;
+    }
+    return NON_EXISTENT_PORT;
+  }
+
+  private void startTunnelThread() {
+    thread = new Thread(new Dispatcher(), "Tunnel Listener");
+    thread.start();
+  }
+
+  public boolean isTunnelThreadAlive() {
+    return (thread != null && thread.isAlive());
+  }
+
+  private class Dispatcher implements Runnable {
```

There is no guarantee because it depends on the implementation of the internal JDBC driver being used, but it is highly likely assuming a sane implementation. It is not possible for the ReadWriteHandler to drop a few bytes and continue executing, because if there's an IOException, the socket is closed and the Tunnel thread exits. This should cause the JDBC driver on the other end to also throw some exception as it fails to read/write to the Tunnel socket.

Now if there is a non-IOException, that could be a problem, since the Tunnel thread will then exit without closing the sockets. In my testing, the MySql JDBC driver times out pretty quickly if the other end of the socket is closed, but to be more careful, I'll add a block to catch all exceptions and close the sockets to force an exception on the other end.